File: spring-impl/src/main/java/org/example/trees/TreeNode.java
package org.example.trees;

/**
 * Binary tree node — Java Record.
 *
 * {@code int value} compiles to a JVM primitive {@code int} field: no boxing on read/write.
 * Jackson 2.15+ deserialises records via the canonical constructor without
 * {@code @JsonCreator}: missing fields map to Java defaults (0 for int, null for refs),
 * so {@code {}} deserialises to {@code TreeNode(0, null, null)}.
 */
public record TreeNode(
    int value,
    TreeNode left,
    TreeNode right
) {}

File: kotlin-impl/src/main/kotlin/org/example/trees/TreeNode.kt
package org.example.trees

/**
 * Binary tree node.
 *
 * `val value: Int` compiles to a JVM primitive `int` field — no boxing on read/write.
 * Default values + Jackson Kotlin module allow deserialization without @JsonCreator.
 * `{}` deserialises to TreeNode(value=0, left=null, right=null), matching Java record behaviour.
 */
data class TreeNode(
    val value: Int = 0,
    val left: TreeNode? = null,
    val right: TreeNode? = null
)
